app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,437 @@
+     * activity, without any file Uris
+     * */
+    Intent filesIntent = getIntent();
+    ArrayList<Uri> fileUriArrayList;
+    fileUriArrayList = filesIntent.getParcelableArrayListExtra(Intent.EXTRA_STREAM);
+    isFileSender = (fileUriArrayList != null && fileUriArrayList.size() > 0);
+
+    setSupportActionBar(actionBar);
+    actionBar.setNavigationIcon(R.drawable.ic_close_white_24dp);
+    actionBar.setNavigationOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View v) {
+        finish();
+      }
+    });
+
+    listViewPeerDevices.setAdapter(
+        new WifiPeerListAdapter(this, R.layout.row_peer_device, availablePeerDevices));
+
+    if (isFileSender) {
+      for (int i = 0; i < fileUriArrayList.size(); i++) {
+        filesForTransfer.add(new FileItem(getFileName(fileUriArrayList.get(i)), TO_BE_SENT));
+      }
+
+      displayFileTransferProgress(filesForTransfer);
+    }
+
+    wifiDirectManager.createWifiDirectManager(sharedPreferenceUtil, alertDialogShower, filesForTransfer, fileUriArrayList);
+  }
+
+  @OnItemClick(R.id.list_peer_devices)
+  void onItemClick(int position) {
+    /* Connection can only be initiated by user of the sender device, & only when transfer has not been started */
+    if (!isFileSender || hasSenderStartedConnection) {
+      return;
+    }
+
+    WifiP2pDevice senderSelectedPeerDevice =
+        (WifiP2pDevice) listViewPeerDevices.getAdapter().getItem(position);
+    wifiDirectManager.sendToDevice(senderSelectedPeerDevice);
+  }
+
+  @Override
+  public boolean onCreateOptionsMenu(@NonNull Menu menu) {
+    getMenuInflater().inflate(R.menu.wifi_file_share_items, menu);
+    return true;
+  }
+
+  @Override
+  public boolean onOptionsItemSelected(@NonNull MenuItem item) {
+    if (item.getItemId() == R.id.menu_item_search_devices) {
+
+      /* Permissions essential for this module */
+      if (!checkCoarseLocationAccessPermission()) {
+        return true;
+      }
+
+      if (!checkExternalStorageWritePermission()) {
+        return true;
+      }
+
+      /* Initiate discovery */
+      if (!wifiDirectManager.isWifiP2pEnabled()) {
+        requestEnableWifiP2pServices();
+        return true;
+      }
+
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && !isLocationServicesEnabled()) {
+        requestEnableLocationServices();
+        return true;
+      }
+
+      showPeerDiscoveryProgressBar();
+      wifiDirectManager.discoverPeerDevices();
+
+      return true;
+    } else {
+      return super.onOptionsItemSelected(item);
+    }
+  }
+
+  @Override
+  public void onUserDeviceDetailsAvailable(@Nullable WifiP2pDevice userDevice) { // Update UI with user device's details
+    if (userDevice != null) {
+      deviceName.setText(userDevice.deviceName);
+      Log.d(TAG, getDeviceStatus(userDevice.status));
+    }
+  }
+
+  @Override
+  public void clearListOfAvailablePeers() {
```

Comment:
this doesn't seem like an appropriate name for this callback, it is commanding not reporting. A callback is an event that occurred, it is up to the implementation to decide what to do about that event.
I think `onConnectionLost` is probably best?

Reply:
I [went with](https://github.com/kiwix/kiwix-android/pull/1276/commits/f071ad60b7716d7c548a5d2a13f510c45482d01d#diff-7e01f6f425daaca67175b74c839c6af2R474) `onConnectionToPeersLost()` - more verbose given the context